File: frontend/src/api/reports.ts
import type { Claim } from './claims'
import { apiClient } from '../lib/apiClient'

export interface ClaimSummaryRow {
  claim_id: number
  employee_id: string
  employee_name: string
  child_id: string
  child_name: string
  invoice_date: string
  invoice_number: string
  invoice_amount: string
  claim_amount: string
  claim_status: Claim['claim_status']
  submitted_date: string | null
  approved_date: string | null
}

export interface ClaimsSummaryResponse {
  rows: ClaimSummaryRow[]
  totals: {
    total_claims: number
    count_by_status: Record<string, number>
    total_invoice_amount: string
    total_approved_amount: string
  }
}

export interface EligibilityUtilizationRow {
  eligibility_id: number
  employee_id: string
  employee_name: string
  child_id: string
  child_name: string
  financial_year: string
  eligible_months: number
  monthly_benefit_amount: string
  allotted_amount: string
  in_progress_amount: string
  approved_amount: string
  remaining_after_approved: string
}

export interface EligibilityUtilizationResponse {
  rows: EligibilityUtilizationRow[]
  totals: {
    total_allotted_amount: string
    total_in_progress_amount: string
    total_approved_amount: string
  }
}

export interface HeadcountResponse {
  total_employees_with_children: number
  total_children: number
  employees_with_one_child: number
  employees_with_two_children: number
  by_financial_year: { financial_year: string; child_count: number }[]
  by_age_bracket: { age_bracket: string; child_count: number }[]
}

export interface PayoutReportRow {
  employee_id: string
  employee_name: string
  child_id: string
  child_name: string
  child_dob: string
  financial_year: string
  apr: string
  may: string
  jun: string
  jul: string
  aug: string
  sep: string
  oct: string
  nov: string
  dec: string
  jan: string
  feb: string
  mar: string
  total_payout: string
}

export interface PayoutReportResponse {
  rows: PayoutReportRow[]
  totals: {
    total_payout: string
  }
}

export interface ClaimsSummaryFilters {
  date_from?: string
  date_to?: string
  status?: string
  employee_id?: string
}

export interface EligibilityUtilizationFilters {
  financial_year?: string
  employee_id?: string
}

export interface PayoutReportFilters {
  financial_year?: string
  employee_id?: string
  child_id?: string
}

export async function getClaimsSummary(
  filters: ClaimsSummaryFilters,
): Promise<ClaimsSummaryResponse> {
  const { data } = await apiClient.get<ClaimsSummaryResponse>('/hr/reports/claims-summary', {
    params: filters,
  })
  return data
}

export async function getEligibilityUtilization(
  filters: EligibilityUtilizationFilters,
): Promise<EligibilityUtilizationResponse> {
  const { data } = await apiClient.get<EligibilityUtilizationResponse>(
    '/hr/reports/eligibility-utilization',
    { params: filters },
  )
  return data
}

export async function getHeadcount(): Promise<HeadcountResponse> {
  const { data } = await apiClient.get<HeadcountResponse>('/hr/reports/headcount')
  return data
}

/** Every financial year that has data, most recent first — feeds the
 * reports' financial-year filter dropdown. */
export async function getFinancialYears(): Promise<string[]> {
  const { data } = await apiClient.get<string[]>('/hr/financial-years')
  return data
}

export async function getPayoutReport(filters: PayoutReportFilters): Promise<PayoutReportResponse> {
  const { data } = await apiClient.get<PayoutReportResponse>('/hr/reports/payout', {
    params: filters,
  })
  return data
}

/** The child's first-13-months auto-paid amounts only, with no claim
 * involved — the sibling of getPayoutReport's claim-driven payout. */
export async function getFirstYearPayoutReport(
  filters: PayoutReportFilters,
): Promise<PayoutReportResponse> {
  const { data } = await apiClient.get<PayoutReportResponse>('/hr/reports/payout/first-year', {
    params: filters,
  })
  return data
}

export async function downloadReportCsv(
  path: string,
  params: Record<string, string | undefined>,
  filename: string,
): Promise<void> {
  const response = await apiClient.get(path, {
    params: { ...params, format: 'csv' },
    responseType: 'blob',
  })
  const url = window.URL.createObjectURL(response.data as Blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  document.body.appendChild(link)
  link.click()
  link.remove()
  window.URL.revokeObjectURL(url)
}

File: backend/app/schemas/reports.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class ClaimSummaryRow(BaseModel):
    claim_id: int
    employee_id: str
    employee_name: str
    child_id: str
    child_name: str
    invoice_date: date
    invoice_number: str
    invoice_amount: Decimal
    claim_amount: Decimal
    institution_name: str | None
    from_date: date | None
    to_date: date | None
    claim_status: str
    submitted_date: datetime | None
    approved_date: datetime | None


class ClaimsSummaryTotals(BaseModel):
    total_claims: int
    count_by_status: dict[str, int]
    total_invoice_amount: Decimal
    total_approved_amount: Decimal


class ClaimsSummaryResponse(BaseModel):
    rows: list[ClaimSummaryRow]
    totals: ClaimsSummaryTotals


class EligibilityUtilizationRow(BaseModel):
    eligibility_id: int
    employee_id: str
    employee_name: str
    child_id: str
    child_name: str
    financial_year: str
    eligible_months: int
    monthly_benefit_amount: Decimal
    allotted_amount: Decimal
    # Computed live from claim/history data for display — see
    # DECISIONS_LOG.md item 32. Not read from EligibilityMaster's own
    # (never-updated) balance columns.
    in_progress_amount: Decimal
    approved_amount: Decimal
    # The child's first-13-months auto-paid total (Childcare_
    # PayoutMonthlyLedger) — shown as its own column so it's never just
    # an invisible gap between allotted_amount and remaining_after_
    # approved (user-caught 2026-09-24: "approved is showing 0, it
    # should have been 56k" for a child with only first-year payout and
    # no claims at all).
    first_year_payout_amount: Decimal
    remaining_after_approved: Decimal


class EligibilityUtilizationTotals(BaseModel):
    total_allotted_amount: Decimal
    total_in_progress_amount: Decimal
    total_approved_amount: Decimal
    total_first_year_payout_amount: Decimal


class EligibilityUtilizationResponse(BaseModel):
    rows: list[EligibilityUtilizationRow]
    totals: EligibilityUtilizationTotals


class HeadcountByFinancialYear(BaseModel):
    financial_year: str
    child_count: int


class HeadcountByAgeBracket(BaseModel):
    age_bracket: str
    child_count: int


class HeadcountResponse(BaseModel):
    total_employees_with_children: int
    total_children: int
    employees_with_one_child: int
    employees_with_two_children: int
    # Grouped by each child's original enrollment financial year — see
    # DECISIONS_LOG.md item 34 for why this isn't a "current status".
    by_financial_year: list[HeadcountByFinancialYear]
    by_age_bracket: list[HeadcountByAgeBracket]


class PayoutReportRow(BaseModel):
    """One Employee + Child + Financial Year, April-through-March — see
    PAYOUT_REQUIREMENTS.md §9's originally-proposed Childcare_
    PayoutSummary shape. Computed live from Childcare_PayoutMonthlyLedger
    (see DECISIONS_LOG.md item 50) rather than a separate stored summary
    table, the same "live report, not a stored aggregate" pattern already
    used for the other HR reports."""

    employee_id: str
    employee_name: str
    child_id: str
    child_name: str
    child_dob: date
    financial_year: str
    apr: Decimal
    may: Decimal
    jun: Decimal
    jul: Decimal
    aug: Decimal
    sep: Decimal
    oct: Decimal
    nov: Decimal
    dec: Decimal
    jan: Decimal
    feb: Decimal
    mar: Decimal
    total_payout: Decimal


class PayoutReportTotals(BaseModel):
    total_payout: Decimal


class PayoutReportResponse(BaseModel):
    rows: list[PayoutReportRow]
    totals: PayoutReportTotals
